lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py

```diff
@@ -0,0 +1,221 @@
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+""" Tests to support utilities."""
+
+import unittest
+
+from cf_units import Unit
+import iris
+from iris.cube import Cube
+from iris.coords import AuxCoord, CellMethod, DimCoord
+from iris.exceptions import InvalidCubeError
+from iris.tests import IrisTest
+import numpy as np
+from numpy import ma
+
+from improver.utilities.spatial import (
+    DiscreteDifferenceBetweenAdjacentGridSquares)
+
+
+def set_up_cube(data, phenomenon_standard_name, phenomenon_units,
+                realizations=np.array([0]), timesteps=1,
+                y_dimension_length=3, x_dimension_length=3):
+    """Create a cube containing multiple realizations."""
+    cube = Cube(data, standard_name=phenomenon_standard_name,
+                units=phenomenon_units)
+    cube.add_aux_coord(AuxCoord(realizations, 'realization',
+                                units='1'))
+    time_origin = "hours since 1970-01-01 00:00:00"
+    calendar = "gregorian"
+    tunit = Unit(time_origin, calendar)
+    cube.add_aux_coord(AuxCoord(np.linspace(402192.5, 402292.5, timesteps),
+                                "time", units=tunit))
+    cube.add_dim_coord(DimCoord(np.linspace(0, 10000, y_dimension_length),
+                                'projection_y_coordinate', units='m'), 0)
+    cube.add_dim_coord(DimCoord(np.linspace(0, 10000, x_dimension_length),
+                                'projection_x_coordinate', units='m'), 1)
+    return cube
+
+
+class Test_create_discrete_difference_cube(IrisTest):
+
+    """Test the create_discrete_difference_cube method."""
+
+    def setUp(self):
+        """Set up cube."""
+        data = np.array([[1, 2, 3],
+                         [2, 4, 6],
+                         [5, 10, 15]])
+        self.cube = set_up_cube(data, "wind_speed", "m s-1")
+        self.plugin = DiscreteDifferenceBetweenAdjacentGridSquares()
+
+    def test_y_dimension(self):
```

Comment:
I don't agree here. Which unit test in particular do you mean? I think the tests here are correctly just testing the code in the method they relate to.
* `Test_calculate_discrete_difference` is only testing the calculation of the difference field: it passes data into the method and checks output matches expectation for four sensible cases.
* `Test_create_discrete_difference_cube` only tests the creation of a cube from the difference array: new cube will have some new metadata and coordinates compared to the original cube. We also want to check data to ensure it hasn't been corrupted by modifying the coordinates for instance.
* `Test_process` just tests the full plugin together, and exceptions. I think this is important to check that both x, and y difference cubes are created and that nothing weird happens in process() around the other two methods.